x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4HttpServerTransportTests.java

```diff
@@ -237,8 +253,97 @@ public void testNoExceptionWhenConfiguredWithoutSslKeySSLDisabled() throws Excep
             Tracer.NOOP,
             new TLSConfig(sslService.getHttpTransportSSLConfiguration(), sslService::createSSLEngine),
             null,
-            randomFrom(Netty4HttpHeaderValidator.NOOP_VALIDATOR, null)
+            randomFrom(VALIDATE_EVERYTHING_VALIDATOR, null)
         );
         assertNotNull(transport.configureServerChannelHandler());
     }
+
+    public void testHttpHeaderAuthenticationErrors() throws Exception {
+        final Settings settings = Settings.builder().put(env.settings()).build();
+        final ThreadPool testThreadPool = new TestThreadPool(TEST_MOCK_TRANSPORT_THREAD_PREFIX);
+        try (
+            Netty4HttpServerTransport transport = Security.getHttpServerTransportWithHeadersValidator(
+                settings,
+                new NetworkService(List.of()),
+                testThreadPool,
+                xContentRegistry(),
+                new NullDispatcher(),
+                randomClusterSettings(),
+                new SharedGroupFactory(settings),
+                Tracer.NOOP,
+                TLSConfig.noTLS(),
+                null,
+                VALIDATE_EVERYTHING_VALIDATOR
+            )
+        ) {
+            final ChannelHandler handler = transport.configureServerChannelHandler();
+            final EmbeddedChannel ch = new EmbeddedChannel(handler);
+            // remove these pipeline handlers as they interfere in the test scenario
+            for (String pipelineHandlerName : ch.pipeline().names()) {
+                if (pipelineHandlerName.equals("decoder")
+                    || pipelineHandlerName.equals("header_validator")
+                    || pipelineHandlerName.equals("encoder")
+                    || pipelineHandlerName.equals("encoder_compress")
+                    || pipelineHandlerName.equals("chunked_writer")) {
+                    ch.pipeline().remove(pipelineHandlerName);
+                }
+            }
+            // this tests a request that cannot be authenticated, but somehow passed authentication
+            // this is the case of an erroneous internal state
+            var writeFuture = testThreadPool.generic().submit(() -> {
+                ch.writeInbound(new DefaultFullHttpRequest(HTTP_1_1, HttpMethod.GET, "/unauthenticable_request"));
+                ch.flushInbound();
+            });
+            writeFuture.get();
+            ch.flushOutbound();
+            Netty4HttpResponse response = ch.readOutbound();
+            assertThat(response.status(), is(HttpResponseStatus.INTERNAL_SERVER_ERROR));
```

Comment:
Why is this a 500 error ?

Reply:
The condition tested here should not normally happen.
The only way it might, is after netty4 upgrades or refactorings that lose the authenticated context. That's way it's a 500, it's an unexpected internal server error.
Because of the relatively loose coupling between the code doing the authn and the code restoring the authn context, I think it's important to test that requests get rejected if somehow the authenticated context is lost before trying to restore it.